compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp

```diff
@@ -0,0 +1,158 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+//=== ReconcileTranslationInfo.cpp ---------------------------------------===//
+//
+// While lowering executable target, the pipelines used are run at a
+// func-like op granularity. Each of these func-like operations set the
+// workgroup size, and subgroup size as required (as part of the
+// `TranslationInfo`). Eventually these have to be reconciled and set
+// appropriately on the surrounding HAL ops for the host runtime to pick them
+// up. In case of inconsistencies, this pass will throw an error.
+//===---------------------------------------------------------------------===//
+
+#include "iree/compiler/Codegen/Common/PassDetail.h"
+#include "iree/compiler/Codegen/Common/Passes.h"
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h"
+
+namespace mlir::iree_compiler {
+
+namespace {
+
+class ReconcileTranslationInfoPass
+    : public ReconcileTranslationInfoBase<ReconcileTranslationInfoPass> {
+public:
+  void runOnOperation() override;
+};
+} // namespace
+
+// Reconcile workgroup sizes across all translation infos.
+static FailureOr<SmallVector<int64_t>> reconcileWorkgroupSize(
+    ArrayRef<IREE::Codegen::TranslationInfoAttr> translationInfos) {
+  if (translationInfos.empty()) {
+    return SmallVector<int64_t>{};
+  }
+  SmallVector<int64_t> reconciledWorkgroupSize =
+      llvm::to_vector(translationInfos.front().getWorkgroupSize());
+  for (auto translationInfo : translationInfos.drop_front()) {
+    auto workGroupSize = llvm::to_vector(translationInfo.getWorkgroupSize());
+    if (workGroupSize != reconciledWorkgroupSize) {
+      return failure();
+    }
+  }
+  return reconciledWorkgroupSize;
+}
+
+// Reconcile subgroup size across all translation infos.
+static FailureOr<int64_t> reconcileSubgroupSize(
+    ArrayRef<IREE::Codegen::TranslationInfoAttr> translationInfos) {
+  if (translationInfos.empty()) {
+    return int64_t();
```

Comment:
I was just trying to see what the tablegen generated code does for `DefaultValuedAttr` and did the same to check for / return default values.

Reply:
> (This is interesting to me because this is my first time seeing it. I'm 99% sure that it is 0. I did run a test locally to verify it. :P )

it is, the umbrella term is value initialization https://en.cppreference.com/w/cpp/language/value_initialization, which ends up being zero-initialization for scalars.